src/annotator/anchoring/trim-range.ts

```diff
@@ -0,0 +1,223 @@
+/**
+ * From which direction to evaluate strings or nodes: from the START of a string
+ * or range looking forward, or from the END of a string or range looking
+ * backward.
+ */
+enum FromPosition {
+  START = 1,
+  END,
+}
+
+/**
+ * Return the offset of the nearest non-whitespace character to `baseOffset`
+ * within `text`, looking in the `direction` indicated. Return -1 if no
+ * non-whitespace character exists between `baseOffset` and the terminus of the
+ * string (start or end depending on `direction`).
+ */
+function trimTextOffset(
+  text: string,
+  baseOffset: number,
+  direction = FromPosition.START
+): number {
+  const nextChar =
+    direction === FromPosition.START ? baseOffset : baseOffset - 1;
+  if (text.charAt(nextChar).trim() !== '') {
+    // baseOffset is already a valid offset
+    return baseOffset;
+  }
+
+  let availableChars: string;
+  let availableNonWhitespaceChars: string;
+
+  if (direction === FromPosition.END) {
+    availableChars = text.substring(0, baseOffset);
+    availableNonWhitespaceChars = availableChars.trimEnd();
+  } else {
+    availableChars = text.substring(baseOffset);
+    availableNonWhitespaceChars = availableChars.trimStart();
+  }
+
+  if (!availableNonWhitespaceChars.length) {
+    return -1;
+  }
+
+  const offsetDelta =
+    availableChars.length - availableNonWhitespaceChars.length;
+
+  return direction === FromPosition.END
+    ? baseOffset - offsetDelta
+    : baseOffset + offsetDelta;
+}
+
+type TrimTextContainerOptions = {
+  direction?: FromPosition;
+  root?: Node;
+};
+
+/**
+ * Return a Node and numerical offset representing the nearest non-whitespace
+ * character to `container`, moving toward `boundaryContainer`.
+ *
+ * Iterate through text nodes in the `direction` indicated toward
+ * `boundaryContainer` from `container`, looking for the first text node that
+ * has non-whitespace text content in it. Return that node, and an offset that
+ * references either the first non-whitespace character (when direction is
+ * FromPosition.START) or the last (direction is FromPosition.END) within that
+ * node.
+ *
+ * @throws {RangeError} If no text node with non-whitespace characters found
+ * between `container` and `boundaryContainer`
+ */
+function trimTextContainer(
+  container: Text,
+  boundaryContainer: Text,
+  { direction = FromPosition.START, root }: TrimTextContainerOptions
+) {
```

Comment:
There are some invariants here about the relationship between `container`, `boundaryContainer` and `root`. If I've got this right:

- `container` and `boundaryContainer` are both descendants of `root`
- `container` occurs before `boundaryContainer`?
- `container` is not `boundaryContainer`?

It might be wise to either assert or at least explicitly state these.